test/circuit/v2/hop.spec.ts

```diff
@@ -0,0 +1,741 @@
+      components.upgrader = new DefaultUpgrader(components, {
+        connectionEncryption: [],
+        muxers: [],
+        inboundUpgradeTimeout: 10000
+      })
+    })
+
+    it('should connect - data limit - src to dest', async () => {
+      const hasReservationStub = sinon.stub(reservationStore, 'hasReservation')
+      const getReservationStub = sinon.stub(reservationStore, 'get')
+      hasReservationStub.resolves(true)
+      getReservationStub.resolves({
+        expire: new Date(Date.now() + 2 * 60 * 1000),
+        addr: multiaddr('/ip4/0.0.0.0'),
+        // set limit
+        limit: {
+          data: BigInt(5),
+          duration: 0
+        }
+      })
+      const dstConn = mockConnection(
+        mockMultiaddrConnection(pair<Uint8Array>(), dstPeer)
+      )
+      const [dstServer, dstClient] = duplexPair<any>()
+      const [srcServer, srcClient] = duplexPair<any>()
+
+      // resolve the destination stream for the server
+      const streamStub = sinon.stub(dstConn, 'newStream')
+      streamStub.resolves(mockStream(dstServer))
+
+      const stub = sinon.stub(components.connectionManager, 'getConnections')
+      stub.returns([dstConn])
+      const srcServerStream = mockStream(srcServer)
+      const srcServerAbort = sinon.spy(srcServerStream, 'abort')
+      const handleHop = expect(handleHopProtocol({
+        connection: conn,
+        stream: pbStream(srcServerStream),
+        request: {
+          type: HopMessage.Type.CONNECT,
+          peer: {
+            id: dstPeer.toBytes(),
+            addrs: []
+          }
+        },
+        relayPeer: relayPeer,
+        relayAddrs: [],
+        reservationStore,
+        peerStore: components.peerStore,
+        connectionManager: components.connectionManager
+      })).to.eventually.fulfilled()
+
+      const dstClientPbStream = pbStream(dstClient)
+      const stopConnectRequest = await dstClientPbStream.pb(StopMessage).read()
+      expect(stopConnectRequest.type).to.eq(StopMessage.Type.CONNECT)
+      // write response
+      dstClientPbStream.pb(StopMessage).write({
+        type: StopMessage.Type.STATUS,
+        status: Status.OK
+      })
+
+      await handleHop
+      const srcClientPbStream = pbStream(srcClient)
+      const response = await srcClientPbStream.pb(HopMessage).read()
+      expect(response.type).to.be.equal(HopMessage.Type.STATUS)
+      expect(response.status).to.be.equal(Status.OK)
+
+      const sourceStream = srcClientPbStream.unwrap()
+      const destStream = dstClientPbStream.unwrap()
+
+      // source to dest, write 4 bytes
+      const sender = pushable()
+      void pipe(sender, sourceStream)
+      sender.push(uint8arrayFromString('01234'))
+      // source to dest, exceed stream limit
+      sender.push(uint8arrayFromString('extra'))
+      const data = await all(destStream.source)
+      expect(data).to.have.length(1)
+      expect(data[0]).to.have.length(5)
+      expect(srcServerAbort.callCount).to.equal(1)
```

Comment:
This sends five bytes, not four.  When four bytes are sent the test fails.

The dialler should send `0123` then `extra`, then the relay should relay five bytes before reset the stream.

Also, the test needs to assert that both streams were reset - from the dialler to the relay and from the relay to the listener.

Reply:
Fixed. I was resetting the stream if the sender attempted to send more than the limit.

> Also, the test needs to assert that both streams were reset - from the dialler to the relay and from the relay to the listener.

The relay enforces the limit, so the abort is called from the relay to the dialler and from the relay to the listener. This is checked.